Fix bad pattern match during handoff with write_once buckets

fadushin opened the pull request:
This PR addresses a bug with write_once buckets that may arise during handoff.  If data is coming into a write_once bucket while handoff is taking place, the riak_core_vnode will crash with a bad_match.

C.f., https://github.com/basho/riak_kv/issues/1138

The fix is to pattern match cleanly against possible return values from handle_command with a KV_W1C_PUT_REQ as input.

A corresponding fix to the verify_handoff_write_once riak_test is on the bugfix/fd/RIAK-1850 branch in the riak_test repo, under PR https://github.com/basho/riak_test/pull/812

fadushin committed:
Enhanced readability

fadushin committed:
Fixed pattern match in handle_handoff_command.

Still had debugging statements in.  Remove before merge.

fadushin committed:
Removed debug logging.

fadushin committed:
Merge branch '2.1' into bugfix/fd/RIAK-1850

JeetKunDoug commented on src/riak_kv_vnode.erl:
```diff
@@ -1046,8 +1046,11 @@ handle_handoff_command(Req=?KV_PUT_REQ{}, Sender, State) ->
-    {forward, NewState};
+    NewState0 = case handle_command(Request, Sender, State) of
+        {noreply, NewState} -> NewState;
+        {reply, _Reply, NewState} -> NewState
```
After talking with Jon, in the {reply, ...} case we should send the reply ourselves with riak_core_vnode:reply/? and then continue to forward the original put request to the new vnode. Otherwise, the current owner will not respond in the non-async case.

fadushin committed:
Modified handle_handoff_command to reply to the sender, if the call to handle_command was going to conter-factually return a reply.

Also snuck in the ability to disable asynchronous puts, via config, instead of only relying on which back end is configured.

fadushin committed:
Merge commit 'a645e25f461bab5eae4227b965180db22f6e3030' into bugfix/fd/RIAK-1850

JeetKunDoug commented:
:+1: e56e471

fadushin commented:
@borshop merge
